uptime.1
.\"             -*-Nroff-*-
.\"
.TH UPTIME "1" "December 2012" "procps-ng" "User Commands"
.SH NAME
uptime \- Tell how long the system has been running.
.SH SYNOPSIS
.B uptime
[\fIoptions\fR]
.SH DESCRIPTION
.B uptime
gives a one line display of the following information.  The current time, how
long the system has been running, how many users are currently logged on, and
the system load averages for the past 1, 5, and 15 minutes.
.PP
This is the same information contained in the header line displayed by
.BR w (1).
.PP
System load averages is the average number of processes that are either in a
runnable or uninterruptable state.  A process in a runnable state is either
using the CPU or waiting to use the CPU.  A process in uninterruptable state
is waiting for some I/O access, eg waiting for disk.  The averages are taken
over the three time intervals.  Load averages are not normalized for the
number of CPUs in a system, so a load average of 1 means a single CPU system
is loaded all the time while on a 4 CPU system it means it was idle 75% of
the time.
.SH OPTIONS
.TP
\fB\-p\fR, \fB\-\-pretty\fR
show uptime in pretty format
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help text
.TP
\fB\-s\fR, \fB\-\-since\fR
system up since, in yyyy-mm-dd MM:HH:SS format
.TP
\fB\-V\fR, \fB\-\-version\fR
display version information and exit
.SH FILES
.TP
.I /var/run/utmp
information about who is currently logged on
.TP
.I /proc
process information
.SH AUTHORS
.B uptime
was written by
.UR greenfie@gauss.\:rutgers.\:edu
Larry Greenfield
.UE
and
.UR johnsonm@sunsite.\:unc.\:edu
Michael K. Johnson
.UE
.SH "SEE ALSO"
.BR ps (1),
.BR top (1),
.BR utmp (5),
.BR w (1)
.SH "REPORTING BUGS"
Please send bug reports to
.UR procps@freelists.org
.UE


NEWS
procps-ng-3.3.8
---------------
  * top adds extensions for linux NUMA/Node capability
  * top has reverted to former %CPU and %MEM precision
    but also provides a '--enable-wide-percent' option
    Debian #707648
  * eliminate the potential library segmentation fault
    Debian #706259, RedHat #951391
  * top now accomodates a window manager like 'screen'
    RedHat #962022
  * ps shows systemd unit and user unit for a process


procps-ng-3.3.7
---------------

  * top adds a powerful new filter feature wherein any
    window can include or exlude selected fields which
    contain specific values - Debian #682082 & #682083
  * top preserves user input for later recall and edit
  * top provides true input editing vs. just backspace
  * top user filtering with exclusion - Debian #682086
  * top field management resize errors no longer fatal
  * top is more responsive to window resize (SIGWINCH)
  * top will now preserve user's context when signaled


procps-ng-3.3.6
---------------

  * top adds a powerful new feature where the contents
    of some file or output from a program, pipeline or
    shell script can be viewed in its own scrollable &
    searchable window while pausing the normal display
  * top highlights matches with 'locate/next' commands
  * top adds summary/task area memory scaling commands
  * top adds a command to suppress insignificant zeros
  * top adds a command line sort field override switch
  * top provides a default PID for kill, nice commands
  * top allows core dump generation, if a signal would
  * pmap adds 2 new options based on smaps, -X and -XX
  * Adjusted pmap pidlist storage Debian #688180
  * programs will permit core-dumps RedHat #87825, 512857
  * receiving a signal is not a crash RedHat #871824, 441656
  * ps: Fixed negative etime RedHat #871819, #433266
  * pkill -c option returns Debian #693783
  * pmap uses correct types for memory allocation Debian #688180
  * expose freeproc, adds function to API Debian #681653


procps-ng-3.3.5
---------------
  * Bumped SONAME to take into acccount API change

procps-ng-3.3.4
---------------
  * Removed ps -aux bogus message
  * w get -i option to display IP addresses
  * watch 8bit fixes Debian #675069
  * Fixed FTBFS for non-linux Debian #677055
  * pkill -u <uid> doesn't need space Debian #676239
  * top enables intra-column horizontal scrolling
  * top can display a task's environment
  * top can display major/minor pg fault deltas, Debian ##84992
  * top provides additional control over column widths
  * top offers user controlled left/right justification

procps-ng-3.3.3
---------------
  * watch -g command repeats until something changes
  * Changed labels of kB etc to KiB Debian #662786
  * top supports hotplugged memory and cpus Debian #351934, #459287
  * top %cpu not distorted when switch to/from threads Debian #441166
  * top can read old toprc configuration files Debian #651213, #651863
  * top supports input paste once again Debian #663334
  * top -p avoids SEGV when pid doesn't exist Debian 668335
  * Various DejaGNU test cases fixed
  * pgrep.1 graoff syntax fix Debian #665425
  * sysctl avoid crash when -p file has unexpected input
  * sysctl .conf off-by-one error Debian #669128
  * libprocps: protect locale env getting overwrite RedHat #548711

  * Documentation catch up (fixed prior to 3.3.3)
    . top repeating "%" wastes space Debian #322984
    . top -U user name parse error Debian #623200
    . top memory leaks Debian #627257
    . top impossible task swap statistics Debian #628462
    . top forest view segfault Debian #650864


procps-ng-3.3.1 --> procps-ng-3.3.2
--------------------------------

  * Redefined library to use version-info
  * NLS added! Programs can use po files to localise the output to suite
    any supported languages.
  * Imported a bunch of distribution patches:
      watch: support unicode
      watch: add precision wait time option -p
      watch: interpret ANSI color code sequences
      watch: add -exec and -beep flags and has better quoting
      w: use environment to set user and from/host column widths
      w: use COLUMNS environment if TIOCGWINSZ fails
      w: bassman emulation with -o option
      vmstat: do not scale si/so just like bi/bo
      libprocps-ng: sysinfo.c: truncate the vmstat figure to 32 bits
      tload: remote unneeded optarg and optind variables
      sysctl: fix up some option processing
      skill: kill prints perror
      skill: do not treat skill null parameter as 0
      skill: fix too greedy option parser
      libprocps-ng: readproc.c: some type conversion help
      ps: rename SZ to SIZE
      ps: add sorting to %mem for ps
      pmap: provide information for -x option
      pgrep: distinguish between invalid commandline parameters and '-?'
      pgrep: fix compiler warning saved_start_time might be used uninitialized
      pgrep: add -c option for counting number of matched proceesses
      pwdx & libprocps-ng: Hurd does not have MAX_PATH defined
      ps: --sort does not work with time argument
      skill: add CR to warning line
      contrib: minimal ps: define mips PAGE_SIZE
      libproc-ng: prettyfy proc mount messages
      ps: add build option to disable ps option warning
      libproc-ng: support building without WCHAR support
      sysctl: remove index() for buildroot
  * top now has a search capability
  * User import is validated better to stop negative values for intervals etc

procps-ng-3.3.0 --> procps-ng-3.3.1
--------------------------------

  * Added DejaGNU regression testing
  * Fixed pgrep -u not finding processes Debian#649591
  * Fixed pgrep crashing
  * vmstat -p <part> finds partitions. Was Debian patch vmstat_part_format
    fixes closed bugs RH#485243 and Debian#588677
  * watch 8-bit clean, Was Debian patch watch_8bitchar
  * slabtop prints plain ASCII in once mode

procps-3.2.8 --> procps-ng-3.3.0
------------------------------------------------------
Debian, Fedora and openSUSE fork of procps.

uptime.c
/*
 * uptime.c - display system uptime
 * Copyright (C) 2012 Craig Small <csmall-procps@enc.com.au>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <errno.h>
#include <getopt.h>
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <sys/time.h>

#include "c.h"
#include "fileutils.h"
#include "nls.h"
#include "proc/sysinfo.h"
#include "proc/whattime.h"
#include "proc/version.h"

static void print_uptime_since()
{
	double now, uptime_secs, idle_secs;
	time_t up_since_secs;
	struct tm *up_since;
	struct timeval tim;

	/* Get the current time and convert it to a double */
	gettimeofday(&tim, NULL);
	now = tim.tv_sec + (tim.tv_usec / 1000000.0);

	/* Get the uptime and calculate when that was */
	uptime(&uptime_secs, &idle_secs);
	up_since_secs = (time_t) ((now - uptime_secs) + 0.5);

	/* Show this */
	up_since = localtime(&up_since_secs);
	printf("%04d-%02d-%02d %02d:%02d:%02d\n",
		up_since->tm_year + 1900, up_since->tm_mon + 1, up_since->tm_mday,
		up_since->tm_hour, up_since->tm_min, up_since->tm_sec);
}

static void __attribute__ ((__noreturn__)) usage(FILE * out)
{
	fputs(USAGE_HEADER, out);
	fprintf(out, _(" %s [options]\n"), program_invocation_short_name);
	fputs(USAGE_OPTIONS, out);
	fputs(_(" -p, --pretty   show uptime in pretty format\n"), out);
	fputs(USAGE_HELP, out);
	fputs(_(" -s, --since    system up since\n"), out);
	fputs(USAGE_VERSION, out);
	fprintf(out, USAGE_MAN_TAIL("uptime(1)"));

	exit(out == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
}

int main(int argc, char **argv)
{
	int c, p = 0;

	static const struct option longopts[] = {
		{"pretty", no_argument, NULL, 'p'},
		{"help", no_argument, NULL, 'h'},
		{"since", no_argument, NULL, 's'},
		{"version", no_argument, NULL, 'V'},
		{NULL, 0, NULL, 0}
	};

#ifdef HAVE_PROGRAM_INVOCATION_NAME
	program_invocation_name = program_invocation_short_name;
#endif
	setlocale (LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	atexit(close_stdout);

	while ((c = getopt_long(argc, argv, "phsV", longopts, NULL)) != -1)
		switch (c) {
		case 'p':
			p = 1;
			break;
		case 'h':
			usage(stdout);
		case 's':
			print_uptime_since();
			return EXIT_SUCCESS;
		case 'V':
			printf(PROCPS_NG_VERSION);
			return EXIT_SUCCESS;
		default:
			usage(stderr);
		}

	print_uptime(p);
	return EXIT_SUCCESS;
}
